url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/expert_system/Key.java

```diff
@@ -0,0 +1,177 @@
+	A_SHARP (10,22,34,46,58,70,82,94,106,120,132,144,156),
+	B		(11,23,35,47,59,71,83,95,107,121,133,145,157);
+	
+	
+	private final int octave0, octave1, octave2, 
+					  octave3, octave4, octave5, 
+					  octave6, octave7, octave8,
+					  octave9, octave10, octave11, octave12;
+	
+	private Key relativeMinor;
+	
+	private static final Key [] allKeys;
+	
+	public int [] majorIntervals = {0,2,2,1,2,2,2};
+	public int [] minorIntervals = {0,2,1,2,2,1,2};
+	
+	private Chord chord;
+	
+	// The allkeys array that stores all the key types for retrieval
+	// in other classes has to be initialise in a static constructor
+	// as until they have been created they cannot be assigned but 
+	// because a static constructor is run at the end of a class 
+	// it makes it possible to initialise the array.
+	static{
+		int i  = 0;
+		allKeys = new Key[24];
+		for (Key k : EnumSet.allOf(Key.class)){
+			
+			allKeys[i]    = k;
+			allKeys[i+12] = k;
+			i++;
+		}
+	}
+	// ##########################################################
+	// #					   CONSTRUCTORS						#
+	// ##########################################################
+	private Key(int oct0, int oct1, int oct2, 
+			    int oct3, int oct4, int oct5,
+			    int oct6, int oct7, int oct8, 
+			    int oct9, int oct10,int oct11, int oct12){
+		
+		
+		this.octave0 = oct0;
+		this.octave1 = oct1;
+		this.octave2 = oct2;
+		this.octave3 = oct3;
+		this.octave4 = oct4;
+		this.octave5 = oct5;
+		this.octave6 = oct6;
+		this.octave7 = oct7;
+		this.octave8 = oct8;
+		this.octave9 = oct9;
+		this.octave10= oct10; 
+		this.octave11 = oct11;
+		this.octave12 = oct12;
+		
+		
+	}
+	// Returns an array containing all the octave fields
+	public int [] getOctaves(){
+		
+		int [] temp = {octave0,octave1,octave2,octave3,
+					   octave4,octave5,octave6,octave7,
+					   octave8,octave9,octave10, octave11, octave12};
+		return temp;
+		
+	}
+	// Returns relativeMinor field
+	public Key getRelativeMinor(){
+		
+		return this.relativeMinor;
+		
+	}
+	// Set the relative minor of each key according to the name
+	// of each enum as using the enum itself doesn't work unless
+	// it has already been initialised so encapsulating this method 
+	// in the constructor is problematic unless done this way.
+	public void setRelativeMinor(){
+		
+		String name = this.name();
+		
+		if(name == "C"){
+		    this.relativeMinor = Key.A;
+		}
+		else if(name == "C_SHARP"){
+			this.relativeMinor = Key.A_SHARP;
+			
+		}
+		else if(name == "D"){
+			this.relativeMinor = Key.B;
+			
+		}
+		else if(name == "D_SHARP"){
+			this.relativeMinor = Key.C;
+		}
+		else if(name == "E"){
+			this.relativeMinor = Key.C_SHARP;
+		}
+		else if(name == "F"){
+			this.relativeMinor = Key.D;
+		}
+		else if(name == "F_SHARP"){
+			this.relativeMinor = Key.D_SHARP;
+		}
+		else if(name == "G"){
+			this.relativeMinor = Key.E;
+		}
+		else if(name == "G_SHARP"){
+			this.relativeMinor = Key.F;
+		}
+		else if(name == "A"){ 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.